### PR TITLE
#164857233 Report article

### PR DESCRIPTION
**Pull request by misocho**
#### What does this PR do?
- Create report article endpoint
- Add tests for report article endpoint

#### Description of Task to be completed?
- Create an endpoint for report article endpoint
- Write tests for report article endpoint
#### Endpoints to be tested
- Post a report
- Flag a report
- List reports
- Get a single report
- Delete a report
- Admin flag an article
- Admin unflag an article
- Admin get flagged articles
#### How should this be manually tested?
- Clone the [repo](https://github.com/rfpremier/ah-django.git)
- Navigate to the project directory
- Branch to `ft/164857233-report-article`
- Create and activate a virtual environment
- Install requirements
```
$ pip install requirements.txt
```
- Make migrations
```
$ python manage.py makemigrations
```
- Run migrations
```
$ python manage.py migrate
``` 
- Run the server
```
$ python manage.py runserver
```
##### On Postman, test the following endpoints
- First signup if you don't have an account `[POST] /api/users/`
Required fields
```
{
            "user": {
                "email": "testuser@gmail.com",
                "username": "testuser",
                "password": "testuserpass123"
            }
}
```
Response
```{
    "user": {
        "email": "testuser@gmail.com",
        "username": "testuser",
        "token": "token"
    }
}
```
- Post an article `[POST] /api/articles/`
- Fill in the authentication token
Required fields
```
{
            "article": {
                "title": "This is the article title",
                "description": "This is the article description",
                "body": "This is the article body",
                "image_url": "https://imageurl.com"
            }

        }
``` 
Response
```
{
    "id": "id",
    "title": "This is the article title",
    "description": "This is the article description",
    "body": "This is the article body",
    "image_url": "https://imageurl.com",
    "author": "author details",
    "created_at": "created at",
    "updated_at": "updated at",
    "slug": "this-is-the-article-title"
}
```
- Post a report `[POST] /api/articles/<article-slug>/report/`
- Fill in the authentication bearer token
 Required fields
```
{
            "report": {
                "report": "This is a sample report"
            }
        }
```
- Response
```
{
    "report": {
        "id": id,
        "article": "article-slug",
        "user": user,
        "report": "This is a sample report",
        "viewed": false,
        "flag": false,
        "action_taken": false,
        "created_at": "created at time",
        "updated_at": "updated at time"
    }
}
```
- Create a superuser
```
$ python manage.py createsuperuser
```
- Login superuser
- Get Admin Bearer token
- List all reports `[GET] /api/articles/report/list/ ` 
- Response
```
{
    "reports": []
}
```
- Get Admin Bearer token
- Get a single report `[GET] /api/articles/report/<report-id>/`
- Response
```
{
        "id": id,
        "article": "article-slug",
        "user": user,
        "report": "This is a sample report",
        "viewed": true,
        "flag": false,
        "action_taken": false,
        "created_at": "created at time",
        "updated_at": "updated at time"
    }
```
- Get Admin Bearer token
- Delete a report `[DELETE] /api/articles/report/<report-id>/action/`
- Response
```
{
    "detail": "Report has been deleted"
}
```
- Flag a report `[POST] /api/articles/report/<report-id>/flag/`
- Requried fields
```
{
	"flag": {
		"flag": "offensive"
	}
}
```
- Response
```
{
    "message": "Article flutter-study-jam-2 has been flagged as offensive"
}
```
- Unflag a report `[DELETE] /api/articles/report/<report-id>/flag/`
- Response
```
{
    "message": "Article flutter-study-jam-2 has been unflagged"
}
```
- Get flagged articles `[GET] /api/articles/report/flag/`
- Response
```
[
    {
        "id": 1,
        "title": "Flutter study jam",
        "description": "Article description",
        "body": "This is the article body",
        "image_url": "http://www.image.com",
        "author": {
            "username": "misocho",
            "bio": "",
            "image": ""
        },
        "created_at": "2019-04-20T10:44:39.985163Z",
        "updated_at": "2019-04-20T11:06:40.965887Z",
        "slug": "flutter-study-jam",
        "avg_rating": 0,
        "rating_count": 0,
        "tags": [
            "tag1",
            "tag2"
        ],
        "favorited": false,
        "favoritesCount": 0,
        "read_time": "1 min read",
        "flag": "offensive"
    }
]
```
#### Background context
- `[GET] /api/articles/` returns only unflagged articles
- `[GET] /api/articles/all/` returns all articles - This endpoint can only be accessed by the admin
- When an article has been flagged the author gets an email notification
- Only the author and the admin can access a flagged article `[GET] /api/articles/flagged-article/`
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164857233

### CODE REVIEW

**Comment by bencyn** [comment]
I tried deleting an article report. Am getting the attached response 
image

**Review by Kimaiyo077 — APPROVED** [review]
It's working as stated on my end.

**Comment by bencyn** [comment]
Kindly include the get flagged articles endpoint in your PR description. 
image

**Comment by Tevinthuku** [comment]
> Kindly include the get flagged articles endpoint in your PR description.
> image

Is this in the README :thinking: ?

File: authors/urls.py [hunk 1]
```diff
@@ -27,4 +26,6 @@
     url(r'^api/', include('authors.apps.comments.urls')),
     url(r'^api/', include('authors.apps.profiles.urls')),
     url(r'^api/', include('authors.apps.bookmarks.urls')),
+    url(r'^api/', include('authors.apps.profiles.urls')),
```
Comment (Tevinthuku) [inline authors/urls.py]: Kindly remove 1 `url(r'^api/', include('authors.apps.profiles.urls'))` profile URL namespace there are 2 in the urlpatterns

Comment (misocho) [reply authors/urls.py]: I have made the changes

**Comment by bencyn** [comment]
> > Kindly include the get flagged articles endpoint in your PR description.
> > image
> 
> Is this in the README  ?

No, the endpoint was lacking in the PR description

**Review by Tevinthuku — APPROVED** [review]
LGTM

**Review by bencyn — COMMENTED** [review]
working as expected good job

**Review by martinMutuma — APPROVED** [review]
Works on my end

**Pull request merged**